pkcs11/src/backend/session.rs

```diff
@@ -479,7 +479,13 @@ impl Session {
             .into_iter()
             .filter(|(_, obj)| {
                 if let Some(kind) = requirements.kind {
-                    kind == obj.kind
+                    if (matches!(requirements.kind, Some(ObjectKind::Certificate))
+                        && kind == obj.kind
+                        && requirements.subject.is_some()) {
+                        requirements.subject == obj.subject
+                    } else {
+                        kind == obj.kind
+                    }
```

Comment:
With this you don't need to coy the `subject` in `Object`

```suggestion
                if let Some(kind) = requirements.kind {
                    kind == obj.kind
                        && matches!(kind, ObjectKind::Certificate)
                        && obj.attr(CKA_SUBJECT).map(|attr| attr.as_bytes())
                            == requirements.subject.as_deref()
                } else {
                    true
                }

```

Reply:
I will do the following, so you can extend easily on further requirements support:
```
                if let Some(kind) = requirements.kind {
                    // kind must match
                    if kind != obj.kind {
                        false
                    // extra checks if kind is Cerificate
                    } else if kind == ObjectKind::Certificate {
                        // When Subject is provided as requirement, it must match
                        requirements.cka_subject.is_none() ||
                            obj.attr(CKA_SUBJECT)
                                .map(|attr| attr.as_bytes())
                                    == requirements.cka_subject.as_deref()
                    // On other kinds, no need for extra checks
                    } else {
                        true
                    }
                } else {
                    true
                }
```

Reply:
No, please make it as simple as possible. Avoid extra complexity for future checks that may never exist.